In [1]:
from pathlib import Path

from vllm import LLM, SamplingParams



INFO 07-15 19:53:40 [__init__.py:44] Available plugins for group vllm.platform_plugins:
INFO 07-15 19:53:40 [__init__.py:46] - metal -> vllm_metal:register
INFO 07-15 19:53:40 [__init__.py:49] All plugins in this group will be loaded. Set `VLLM_PLUGINS` to control which plugins to load.


/Users/arjitsharma/Documents/dev/ai_venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


WARNING 07-15 19:53:42 [compat.py:434] Could not install vLLM ByteLevel tokenizer compatibility patch because vLLM tokenizer registry is unavailable: cannot import name 'common_broadcastable_dtype' from partially initialized module 'vllm.utils.torch_utils' (most likely due to a circular import) (/Users/arjitsharma/Documents/dev/ai_venv/lib/python3.12/site-packages/vllm/utils/torch_utils.py)
INFO 07-15 19:53:43 [__init__.py:237] Platform plugin metal is activated
INFO 07-15 19:53:43 [importing.py:88] Triton not installed or not compatible; certain GPU-related functions will not be available.


objc[47344]: Class SDLApplication is implemented in both /Users/arjitsharma/Documents/dev/ai_venv/lib/python3.12/site-packages/cv2/.dylibs/libSDL2-2.0.0.dylib (0x137404890) and /opt/homebrew/Cellar/sdl2/2.32.6/lib/libSDL2-2.0.0.dylib (0x13df43c88). This may cause spurious casting failures and mysterious crashes. One of the duplicates must be removed or renamed.
objc[47344]: Class SDLAppDelegate is implemented in both /Users/arjitsharma/Documents/dev/ai_venv/lib/python3.12/site-packages/cv2/.dylibs/libSDL2-2.0.0.dylib (0x1374048e0) and /opt/homebrew/Cellar/sdl2/2.32.6/lib/libSDL2-2.0.0.dylib (0x13df43cd8). This may cause spurious casting failures and mysterious crashes. One of the duplicates must be removed or renamed.
objc[47344]: Class SDLTranslatorResponder is implemented in both /Users/arjitsharma/Documents/dev/ai_venv/lib/python3.12/site-packages/cv2/.dylibs/libSDL2-2.0.0.dylib (0x137404958) and /opt/homebrew/Cellar/sdl2/2.32.6/lib/libSDL2-2.0.0.dylib (0x13df43d50). This may cause 

In [5]:
MODEL_DIR = str(Path.cwd() / "models" / "TinyLlama-1.1B-Chat-v1.0")
PROMPT = "Explain what an AI agent is, in two sentences."
MAX_TOKENS = 200
TEMPERATURE = 0.7

In [4]:


print(f"Loading local model from: {MODEL_DIR}")
llm = LLM(model=MODEL_DIR)
sampling_params = SamplingParams(temperature=TEMPERATURE, max_tokens=MAX_TOKENS)

# TinyLlama-Chat is instruct-tuned, so use the chat template.
outputs = llm.chat([{"role": "user", "content": PROMPT}], sampling_params)
response = outputs[0].outputs[0].text.strip()

print("\n--- Prompt ---")
print(PROMPT)
print("\n--- Response ---")
print(response)



Loading local model from: /Users/arjitsharma/Documents/dev/GenAIPractice/llm_utils/models/TinyLlama-1.1B-Chat-v1.0
INFO 07-15 19:55:34 [api_utils.py:273] non-default args: {'disable_log_stats': True, 'model': '/Users/arjitsharma/Documents/dev/GenAIPractice/llm_utils/models/TinyLlama-1.1B-Chat-v1.0'}
INFO 07-15 19:55:34 [model.py:619] Resolved architecture: LlamaForCausalLM
INFO 07-15 19:55:34 [model.py:1776] Using max model len 2048
INFO 07-15 19:55:34 [scheduler.py:252] Chunked prefill is enabled with max_num_batched_tokens=8192.
INFO 07-15 19:55:34 [vllm.py:1042] Asynchronous scheduling is enabled.
INFO 07-15 19:55:34 [kernel.py:292] Final IR op priority after setting platform defaults: IrOpPriorityConfig(rms_norm=['native'], fused_add_rms_norm=['native'])
INFO 07-15 19:55:35 [platform.py:625] Metal: chunked prefill enabled (paged attention), max_num_batched_tokens=8192
INFO 07-15 19:55:35 [platform.py:724] Metal memory: 38.7GB total, 20.9GB available
WARNING 07-15 19:55:35 [vllm.py:

objc[47465]: Class SDLApplication is implemented in both /Users/arjitsharma/Documents/dev/ai_venv/lib/python3.12/site-packages/cv2/.dylibs/libSDL2-2.0.0.dylib (0x135334890) and /opt/homebrew/Cellar/sdl2/2.32.6/lib/libSDL2-2.0.0.dylib (0x13c00fc88). This may cause spurious casting failures and mysterious crashes. One of the duplicates must be removed or renamed.
objc[47465]: Class SDLAppDelegate is implemented in both /Users/arjitsharma/Documents/dev/ai_venv/lib/python3.12/site-packages/cv2/.dylibs/libSDL2-2.0.0.dylib (0x1353348e0) and /opt/homebrew/Cellar/sdl2/2.32.6/lib/libSDL2-2.0.0.dylib (0x13c00fcd8). This may cause spurious casting failures and mysterious crashes. One of the duplicates must be removed or renamed.
objc[47465]: Class SDLTranslatorResponder is implemented in both /Users/arjitsharma/Documents/dev/ai_venv/lib/python3.12/site-packages/cv2/.dylibs/libSDL2-2.0.0.dylib (0x135334958) and /opt/homebrew/Cellar/sdl2/2.32.6/lib/libSDL2-2.0.0.dylib (0x13c00fd50). This may cause 

INFO 07-15 19:55:38 [importing.py:88] Triton not installed or not compatible; certain GPU-related functions will not be available.
(EngineCore pid=47465) INFO 07-15 19:55:39 [core.py:114] Initializing a V1 LLM engine (v0.25.1) with config: model='/Users/arjitsharma/Documents/dev/GenAIPractice/llm_utils/models/TinyLlama-1.1B-Chat-v1.0', speculative_config=None, tokenizer='/Users/arjitsharma/Documents/dev/GenAIPractice/llm_utils/models/TinyLlama-1.1B-Chat-v1.0', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.bfloat16, max_seq_len=2048, download_dir=None, load_format=auto, tensor_parallel_size=1, pipeline_parallel_size=1, data_parallel_size=1, decode_context_parallel_size=1, dcp_comm_backend=ag_rs, disable_custom_all_reduce=True, quantization=None, quantization_config=None, enforce_eager=False, enable_return_routed_experts=False, kv_cache_dtype=auto, device_config=cpu, structured_outputs_config=StructuredOutputs

mx.metal.device_info is deprecated and will be removed in a future version. Use mx.device_info instead.


(EngineCore pid=47465) INFO 07-15 19:55:40 [model_lifecycle.py:229] MLX-LM model loaded in 0.78s: /Users/arjitsharma/Documents/dev/GenAIPractice/llm_utils/models/TinyLlama-1.1B-Chat-v1.0
(EngineCore pid=47465) INFO 07-15 19:55:44 [cache_policy.py:748] Paged attention: VLLM_METAL_MEMORY_FRACTION=auto, defaulting to 0.90 for paged path
(EngineCore pid=47465) INFO 07-15 19:55:44 [cache_policy.py:581] Paged attention memory breakdown: metal_limit=30.15GB, fraction=0.9, usable_metal=27.14GB, model_memory=2.20GB, overhead=0.86GB, kv_budget=24.07GB, per_block_bytes=360448, num_blocks=66782, max_tokens_cached=1068512
(EngineCore pid=47465) INFO 07-15 19:55:46 [kv_cache.py:167] KV cache: 24071.4 MB (22 layers, 66782 blocks, 16 tokens/block)
(EngineCore pid=47465) INFO 07-15 19:55:46 [cache_policy.py:611] Paged attention enabled: 22 layers patched, 66782 blocks allocated (block_size=16, mla=False, turboquant=False, k_quant=N/A)
(EngineCore pid=47465) INFO 07-15 19:55:46 [cache_policy.py:652] Pag

Rendering conversations:   0%|                                                                                                                      | 0/1 [00:00<?, ?it/s]

INFO 07-15 19:55:49 [hf.py:548] Detected the chat template content format to be 'string'. You can set `--chat-template-content-format` to override this.


Processed prompts: 100%|██████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.94s/it, est. speed input: 14.44 toks/s, output: 35.06 toks/s]


--- Prompt ---
Explain what an AI agent is, in two sentences.

--- Response ---
An AI agent is a computer program that has the ability to simulate or mimic the behaviors and cognitive processes of human beings, including decision-making, learning, and problem-solving. AI agents can be used in a wide range of applications, from autonomous driving to robotics and healthcare.


--- Prompt ---
Explain what an AI agent is, in two sentences.

--- Response ---
An AI agent is a computer program that has the ability to simulate or mimic the behaviors and cognitive processes of human beings, including decision-making, learning, and problem-solving. AI agents can be used in a wide range of applications, from autonomous driving to robotics and healthcare.

### Explaination: https://claude.ai/code/artifact/985dc5f3-037e-437b-a6e3-e71a279eae2c

It walks through two acts, each tied to real log lines from your terminal:

* **Act I** — Startup (8 steps): plugin discovery → process split (your script and EngineCore are actually two separate OS processes talking over IPC) → config resolution → the dual MLX/PyTorch-mps device registration (root cause of the shutdown error) → weight loading → paged-attention KV-cache sizing (66,782 pages, OS-paging-style memory management) → Metal kernel warm-up → ready.
* **Act II** — Inference (7 steps): chat templating → tokenization → scheduling/page allocation → prefill (whole prompt in one parallel pass) → the autoregressive decode loop (forward pass → sample → append → repeat) → detokenization → return across the process boundary.
* Two diagrams: one showing the driver-process/EngineCore-process/GPU architecture, one showing the prefill-then-decode-loop cycle.

# vLLM
- vLLM is a high-throughput inference engine (**continuous batching, paged attention, GPU kernel optimization**) for serving LLMs, not just a one-off generate() call.
- Why this beats calling LLM(...) per-request: the ~8s boot cost from the trace we just walked through happens once at server startup, not per request; multiple requests get batched together automatically by the scheduler (that's the 521x concurrency headroom you saw); and you get streaming, retries, and a stable API for free instead of hand-rolling them.
- For a Application, vLLM ships a built-in server that keeps the engine alive and exposes an OpenAI-compatible HTTP API - **vllm serve**

This is how you'd build the application:
```mermaid
flowchart LR
    A[Clients<br/>Web UI, API, cURL]
    B[vLLM API Server <br/> vllm serve]
    C[Scheduler<br/>Continuous Batching]
    D[Inference Engine <br/>KV Cache <br/> vllm]
    E[LLM Model]
    F[GPU Cluster]

    A -->|POST /v1/chat/completions| B
    B --> C
    C --> D
    D --> E
    E --> F

    F -->|Generated Tokens| A
```


In [10]:
import subprocess
import sys
import threading

vllm_bin = str(Path(sys.executable).parent / "vllm")
log_path = Path.cwd() / "vllm_serve.log"
log_file = open(log_path, "w")

server_proc = subprocess.Popen(
    [
        vllm_bin, "serve", MODEL_DIR,
        "--served-model-name", "tinyllama",
        "--port", "8000",
    ],
    stdout=subprocess.PIPE,
    stderr=subprocess.STDOUT,
    text=True,
    bufsize=1,
)

ready = threading.Event()

def _stream_until_ready():
    for line in server_proc.stdout:
        log_file.write(line)
        print(line, end="")
        if "Application startup complete" in line or "Uvicorn running on" in line:
            ready.set()
            break

def _drain_to_file():
    # After startup, keep draining stdout to the log file (not the notebook)
    # so the pipe never fills up and blocks the server.
    for line in server_proc.stdout:
        log_file.write(line)

threading.Thread(target=_stream_until_ready, daemon=True).start()
startup_ok = ready.wait(timeout=180)
threading.Thread(target=_drain_to_file, daemon=True).start()

print(f"\n{'Ready' if startup_ok else 'Timed out waiting'} — pid={server_proc.pid}")
print(f"Further logs -> {log_path}")
print("Stop it later with: server_proc.terminate()")

INFO 07-15 20:31:21 [__init__.py:44] Available plugins for group vllm.platform_plugins:
INFO 07-15 20:31:21 [__init__.py:46] - metal -> vllm_metal:register
INFO 07-15 20:31:21 [__init__.py:49] All plugins in this group will be loaded. Set `VLLM_PLUGINS` to control which plugins to load.
WARNING 07-15 20:31:23 [compat.py:434] Could not install vLLM ByteLevel tokenizer compatibility patch because vLLM tokenizer registry is unavailable: cannot import name 'common_broadcastable_dtype' from partially initialized module 'vllm.utils.torch_utils' (most likely due to a circular import) (/Users/arjitsharma/Documents/dev/ai_venv/lib/python3.12/site-packages/vllm/utils/torch_utils.py)
INFO 07-15 20:31:23 [__init__.py:237] Platform plugin metal is activated
INFO 07-15 20:31:24 [importing.py:88] Triton not installed or not compatible; certain GPU-related functions will not be available.
objc[48200]: Class SDLApplication is implemented in both /Users/arjitsharma/Documents/dev/ai_venv/lib/python3.12/s

In [15]:
%%bash
content = 
curl -sS http://localhost:8000/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
      "model": "tinyllama",
      "messages": [{"role": "user", "content": "Explain what an AI agent is, in two sentences."}],
      "max_tokens": 200,
      "temperature": 0.7
    }'

{"id":"chatcmpl-8fd891024125c17a","object":"chat.completion","created":1784127950,"model":"tinyllama","choices":[{"index":0,"message":{"role":"assistant","content":"An AI agent is a software or machine that is designed to simulate intelligent behavior in a virtual environment or real world. It can perceive and respond to its environment, making it able to perform tasks autonomously, adaptively, and frequently. AI agents are often used in various industries, including healthcare, manufacturing, and finance, to improve efficiency, reduce costs, and enhance customer experience.","refusal":null,"annotations":null,"audio":null,"function_call":null,"reasoning":null},"logprobs":null,"finish_reason":"stop","stop_reason":null,"token_ids":null,"routed_experts":null}],"service_tier":null,"system_fingerprint":"vllm-0.25.1-eafbf2a9","usage":{"prompt_tokens":28,"total_tokens":114,"completion_tokens":86,"prompt_tokens_details":null},"prompt_logprobs":null,"prompt_token_ids":null,"prompt_text":null,"k

In [16]:
import requests

content = {
    "model": "tinyllama",
    "messages": [{"role": "user", "content": "Explain what an AI agent is, in two sentences."}],
    "max_tokens": 200,
    "temperature": 0.7,
}

resp = requests.post("http://localhost:8000/v1/chat/completions", json=content)
resp.json()

{'id': 'chatcmpl-9fd3b75471739a00',
 'object': 'chat.completion',
 'created': 1784128207,
 'model': 'tinyllama',
 'choices': [{'index': 0,
   'message': {'role': 'assistant',
    'content': 'An AI agent is a computer program designed to simulate and replicate intelligent behavior and intelligence. It operates according to predetermined rules and algorithms and can learn and adapt over time to achieve its objectives. AI agents are designed to perform a range of tasks, from simple tasks like answering questions and performing basic calculations to more complex tasks like playing games and handling complex situations. They are frequently used in fields such as finance, healthcare, and automotive manufacturing.',
    'refusal': None,
    'annotations': None,
    'audio': None,
    'function_call': None,
    'reasoning': None},
   'logprobs': None,
   'finish_reason': 'stop',
   'stop_reason': None,
   'token_ids': None,
   'routed_experts': None}],
 'service_tier': None,
 'system_fingerpri

In [18]:


import requests

content = {
    "model": "tinyllama",
    "messages": [{"role": "user", "content": " -d SyntaxError: unterminated string literal (detected at line 4)"}],
    "max_tokens": 200,
    "temperature": 0.7,
}

resp = requests.post("http://localhost:8000/v1/chat/completions", json=content)
resp.json()

{'id': 'chatcmpl-9176f2b50217106e',
 'object': 'chat.completion',
 'created': 1784128364,
 'model': 'tinyllama',
 'choices': [{'index': 0,
   'message': {'role': 'assistant',
    'content': 'The `list` function prompts the user to enter a string and converts it into a list using `map()` and `reduce()`. However, after running the program, you receive the error message "SyntaxError: unterminated string literal (detected at line 4)" because the input string is not terminated by a newline character.\n\nTo fix this, you need to add a new newline character at the end of the user input before calling the `reduce()` function. Here\'s a modified version of the program that includes this addition:\n\n```python\nimport sys\n\ndef list_to_str(lst):\n    return "".join(l for l in lst if l.isalnum())\n\ndef list_to_int(lst):\n    return int("".join(x for x in lst if x.isdigit()))\n\ndef list_to_float(lst):\n    return float("".join(x for x',
    'refusal': None,
    'annotations': None,
    'audio':